### PR TITLE
Contest solution submit fixes

### PR DESCRIPTION
**Pull request by BojidarTonev**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/1208

### CODE REVIEW

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 1]
```diff
@@ -178,15 +177,53 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
...
+            int id,
+            string? password,
+            bool? hasConfirmedParticipation,
+            bool isOfficial)
+        {
+            var contest = this.contestsData
+                .GetByIdQuery(id)
+                .Include(c => c.Category)
+                .Include(c => c.ProblemGroups)
+                .ThenInclude(pg => pg.Problems)
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:195]: Why do we need the Problems to validate the user registration for a contest?

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 2]
```diff
@@ -142,8 +142,7 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
             return contestDetailsServiceModel;
         }
 
-        public async Task<RegisterUserForContestServiceModel> RegisterUserForContest(
-            int id, string? password, bool? hasConfirmedParticipation, bool isOfficial)
+        public async Task<RegisterUserForContestServiceModel> RegisterUserForContest(int id, bool isOfficial)
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:145]: Now this method does not do the registration part. It should be renamed.

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 3]
```diff
@@ -178,15 +177,53 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
...
+            var participant = await this.participantsData
+                .GetByContestByUserAndByIsOfficial(
+                    id,
+                    userProfile!.Id,
+                    isOfficial);
+
+            var userIsAdminOrLecturerInContest = await this.lecturersInContestsBusiness.IsCurrentUserAdminOrLecturerInContest(contest?.Id);
+            bool shouldRequirePassword = ShouldRequirePassword(contest!, participant, isOfficial);
+            bool shouldConfirmParticipation =
+                ShouldConfirmParticipation(participant, isOfficial, contest!.IsOnlineExam, userIsAdminOrLecturerInContest);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:221]: Can we reuse the method where this information is retrieved in `RegisterUserForContestServiceModel`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs (hunk 3 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:221]: We do, this change does this so both methods use ShouldRequirePassword ShouldConfirmParticipation

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs (hunk 1 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:195]: We need them for `AssignRandomProblemsToParticipant` logic.

Comment (gogo4ds) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:193]: By looking the name it seems the method does only some kind of validation, but not actual registration.

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 4]
```diff
@@ -177,16 +174,54 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
...
+        {
+            var contest = this.contestsData
+                .GetByIdQuery(id)
+                .Include(c => c.Category)
+                .Include(c => c.ProblemGroups)
+                .ThenInclude(pg => pg.Problems)
+                .FirstOrDefault();
+
+            var user = this.userProviderService.GetCurrentUser();
+            var userProfile = await this.usersBusinessService.GetUserProfileById(user.Id);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:197]: Why do you need this `userProfile`?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-solution-submit/ContestSolutionSubmitPage.tsx [hunk 5]
```diff
@@ -298,9 +304,8 @@ const ContestSolutionSubmitPage = () => {
                 page: selectedSubmissionsPage,
                 isOfficial: isCompete,
             });
-            setUploadedFile(null);
         } catch {
-            setUploadedFile(null);
+            // no need to handle error here since it's handled in rtk error of POST request
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-solution-submit/ContestSolutionSubmitPage.tsx:308]: Why not just remove the try-catch block

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-register/ContestRegister.tsx [hunk 6]
```diff
@@ -39,33 +39,46 @@ const ContestRegister = () => {
...
+    useEffect(() => {
+        if (isLoading || !data) {
+            return;
+        }
+        if (!requirePassword && !shouldConfirmParticipation && !isRegisteredSuccessfully) {
+            // eslint-disable-next-line promise/catch-or-return
             registerUserForContest({
                 id: Number(contestId),
-                isOfficial: participationType === ContestParticipationType.Compete,
+                isOfficial: participationType === 'compete',
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-register/ContestRegister.tsx:56]: Why `ContestParticipationType.Compete` is removed and a string is introduced instead?

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/portals/client/ClientPortal.tsx [hunk 7]
```diff
@@ -90,21 +83,25 @@ const ClientPortal = () => {
...
+            path: '/contests/edit/:contestId',
             Element: ContestEditPage,
         },
         {
             path: '/contests/:contestId/:participationType',
             Element: ContestSolutionSubmitPage,
         },
         {
-            path: '/contests/:contestId/:participationType/results/:resultType',
+            path: '/contest/:contestId/:participationType/results/:resultType',
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/portals/client/ClientPortal.tsx:98]: Are we sure it is `contest` here, and not `contests`?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-solution-submit/ContestSolutionSubmitPage.tsx (hunk 5 above)
Comment (BojidarTonev) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-solution-submit/ContestSolutionSubmitPage.tsx]: done

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-register/ContestRegister.tsx (hunk 6 above)
Comment (BojidarTonev) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/pages/contests/contest-register/ContestRegister.tsx]: fixed

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/portals/client/ClientPortal.tsx (hunk 7 above)
Comment (BojidarTonev) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/portals/client/ClientPortal.tsx]: nice catch, fixed

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**